### ISSUE TITLE
AutoStandby doesn't get activated as expected

### ISSUE THREAD

**Issue by zwim** [koreader/koreader#9157]
* KOReader version: 2022.05.1-14
* Device: Sage

#### Issue
Standby gets activated way too late

#### Steps to reproduce
Disable almost all plugins (except power saving), disable WiFi. Plug in the device and check the current with a USB-tester.

You will see the device enter the first standby as expected. But later standbys come too late.

#### Possible cause

@NiLuJe
https://github.com/koreader/koreader/blob/b26d3e981534377f9ff8f9bddddaac99a366e871/plugins/autosuspend.koplugin/main.lua#L604 seems to delay the `self.leave_standby_task` too long.

##### `crash.log` (if applicable)

Excerpt from crash.log
```
06/01/22-15:12:48 DEBUG AutoSuspend: onAllowStandby
06/01/22-15:12:48 DEBUG AutoSuspend: entering standby with a wakeup alarm in 13 s
06/01/22-15:13:01 DEBUG AutoSuspend: left standby after 12.786388 s   ----- OK
06/01/22-15:13:13 DEBUG AutoSuspend: onLeaveStandby                      ---- to late by 13s
06/01/22-15:13:13 DEBUG AutoSuspend: unschedule suspend/shutdown timer
06/01/22-15:13:13 DEBUG AutoSuspend: start suspend/shutdown timer at 17664.527280
06/01/22-15:13:13 DEBUG AutoSuspend: scheduling next suspend check in 900
06/01/22-15:13:13 DEBUG AutoSuspend: scheduling next shutdown check in 259200
06/01/22-15:13:13 DEBUG AutoSuspend: start standby timer at 17664.527280
06/01/22-15:13:13 DEBUG AutoSuspend: blown deadline, delaying standby
06/01/22-15:13:13 DEBUG AutoSuspend: scheduling next standby check in 4
06/01/22-15:13:13 DEBUG AutoSuspend: preventStandby
06/01/22-15:13:17 DEBUG AutoSuspend: allowStandby
06/01/22-15:13:17 DEBUG AutoSuspend: onAllowStandby
06/01/22-15:13:17 DEBUG AutoSuspend: entering standby with a wakeup alarm in 44 s
06/01/22-15:14:01 DEBUG AutoSuspend: left standby after 43.080423 s  ----- OK
06/01/22-15:14:45 DEBUG AutoSuspend: onLeaveStandby                     ---- to late by 45 seconds
06/01/22-15:14:45 DEBUG AutoSuspend: unschedule suspend/shutdown timer
06/01/22-15:14:45 DEBUG AutoSuspend: start suspend/shutdown timer at 17664.527280
06/01/22-15:14:45 DEBUG AutoSuspend: scheduling next suspend check in 900
06/01/22-15:14:45 DEBUG AutoSuspend: scheduling next shutdown check in 259200
06/01/22-15:14:45 DEBUG AutoSuspend: start standby timer at 17664.527280
06/01/22-15:14:45 DEBUG AutoSuspend: blown deadline, delaying standby
06/01/22-15:14:45 DEBUG AutoSuspend: scheduling next standby check in 4
06/01/22-15:14:45 DEBUG AutoSuspend: preventStandby
06/01/22-15:14:49 DEBUG AutoSuspend: allowStandby
06/01/22-15:14:49 DEBUG AutoSuspend: onAllowStandby
06/01/22-15:14:49 DEBUG AutoSuspend: entering standby with a wakeup alarm in 12 s
06/01/22-15:15:01 DEBUG AutoSuspend: left standby after 11.925828 s   ------ OK
06/01/22-15:15:13 DEBUG AutoSuspend: onLeaveStandby                      ------ to late by 13 secs
06/01/22-15:15:13 DEBUG AutoSuspend: unschedule suspend/shutdown timer
06/01/22-15:15:13 DEBUG AutoSuspend: start suspend/shutdown timer at 17664.527280
```   

[crash.log](https://github.com/koreader/koreader/files/8814976/crash.log)

**Comment by NiLuJe** [comment]
The only way I could reproduce something similiar was in degenerate use-cases, where there's no UI frame after the `tickAfterNext`, so it just blocks on input polling on the first "unwrapping" of the task.

In which case you need an input event to trigger a new UI frame.

(Perfect testcase being a device that can't standby while plugged in, e.g., a H2O in USBNet mode).

----

TL;DR: I'm going to need verbose debug logs to see if there are indeed input events ;o).

**Comment by zwim** [comment]
The attached crash.log should be verbose (there was no tap ).
But even with taps, the problem stays. I will provide a log tomorrow.

Thanks for the hints, maybe I will have some time to look into it (a shoot in the dark: I will try a scheduled dummy right before autostandy.)

**Comment by NiLuJe** [comment]
I've identified the issue, I'll try to push a lightly tested fix tonight ;).

**Comment by NiLuJe** [comment]
c.f., https://github.com/NiLuJe/koreader/tree/uimanager-tickafternext-frame

(Untested on actual standby-aware hardware, WIP, etc, etc.).

**Comment by NiLuJe** [comment]
Okay, looked at the logs, in your case, it was because the wakeup source is the wake alarm.

That's slightly less edge-casey than my own testcase, so, hopefully, that won't even delay the Event anymore in this case, because we might actually go through the task queue in this case ;).

**Comment by KaMyKaSii** [comment]
@zwim sorry for the off-topic question, but how do you know that the current the device is drawing from the USB is for its own operation and not for charging the battery? Was the battery already full when testing?

**Comment by zwim** [comment]
Its not off-topic. If standby works as expected there should be current drop 4sec after a wakeup (tap, footer refresh...) In my case there this does not happen.

You can also see in the logs that autostandby does not start.

**Comment by NiLuJe** [comment]
The comment is sound, though: you need a full battery to clearly see the difference, otherwise you're indeed going to mostly see the power drain needed for the battery top up ;).
